### DISCUSSION TITLE
How to see token values or debug premake scripts?

### DISCUSSION THREAD

**Discussion by ShaiAvr** [premake/premake-core#1983] [Q&A]
I am using premake for my project and one thing I find annoying is that it's hard to debug scripts or print values of tokens.
For example, how can I print the values of `%{cfg.system}` or `%{cfg.architecture}` in order to understand what their values are?

Also, the documentation seems incomplete about tokens. The only page about tokens is [this](https://premake.github.io/docs/Tokens/). It lists a bunch of tokens without further explanations or examples. Let's look at `%{cfg.system}` for example. What values does it get for different platforms? I found in StackOverflow that it can be `windows`, `linux`, or `macosx`. Are there more values such as `ios`, `android` etc?

I'd like to see further explanations about each token and more examples. I'd also like to have a way to print these values or debug premake scripts in general. Is there a way to achieve this that I am missing?
Thanks in advance.

**Comment by Jarod42** [comment]
> how can I print the values of %{cfg.system} or %{cfg.architecture}

Not necessary an ideal solution, but a
```
prebuildmessage "%{cfg.system} or %{cfg.architecture}"
```
would allow to have the message in prebuild step (So either run the given project or investigate on the generated solution)

> Also, the documentation seems incomplete about tokens. 

Most documentation is not complete ;-)

> Let's look at `%{cfg.system}` for example.

I would say value from [`system`](https://premake.github.io/docs/system) api


>  I'd also like to have a way to print these values or debug premake scripts in general. Is there a way to achieve this that I am missing?

I would say take a small project and check the generated solution.
(I made a [samples repo](https://github.com/Jarod42/premake-sample-projects) mostly for integration testing, but it might help if you are stuck with one of the tested api).

Some generator might be more appropriate/correct than others depending of your usage/knowledge.

Common mistakes when debugging are "scope" (`configurations`/`filter`):
- indentation might be misleading, it doesn't reflect the actual scope.
- they are not equivalent to an `if`.
- they should be reset

```
filter "some condition"
  print("some condition hit") -- (always) executed when running script (never in generated solutions)
  buildoptions("specific options")
buildoptions("general options") -- still in the filter, require a `filter {}` to actually be global
```
